File: plugins/motherduck/skills/motherduck-build-cfa-app/artifacts/customer_routing_example.ts
export {};
declare const process: { env: Record<string, string | undefined> };

type EventRow = { event_id: number; event_type: string; revenue: number };

function normalizeMetadataValue(value: string | undefined, fallback: string): string {
  const raw = (value ?? "").trim();
  if (!raw) return fallback;
  const normalized = raw.replace(/[^A-Za-z0-9._-]+/g, "-").replace(/^[-._]+|[-._]+$/g, "");
  return normalized || fallback;
}

function buildUseCaseUserAgent(): string {
  const harness = normalizeMetadataValue(process.env.MOTHERDUCK_AGENT_HARNESS, "unknown");
  const llm = normalizeMetadataValue(process.env.MOTHERDUCK_AGENT_LLM, "unknown");
  return `agent-skills/2.2.2(harness-${harness};llm-${llm})`;
}

function summarizeCustomer(rows: EventRow[]): Array<{ event_type: string; total_revenue: number }> {
  const totals = new Map<string, number>();
  for (const row of rows) {
    totals.set(row.event_type, (totals.get(row.event_type) ?? 0) + row.revenue);
  }
  return Array.from(totals.entries())
    .map(([event_type, total_revenue]) => ({ event_type, total_revenue }))
    .sort((a, b) => b.total_revenue - a.total_revenue);
}

const customerData: Record<string, EventRow[]> = {
  acme: [
    { event_id: 1, event_type: "search", revenue: 12.5 },
    { event_id: 2, event_type: "checkout", revenue: 18.0 },
  ],
  globex: [
    { event_id: 1, event_type: "signup", revenue: 4.0 },
    { event_id: 2, event_type: "invoice_paid", revenue: 9.5 },
  ],
};

const result = {
  backend: {
    mode: "typescript-companion",
    databases: {
      customer_acme: "customer_acme",
      customer_globex: "customer_globex",
    },
    user_agent: buildUseCaseUserAgent(),
  },
  pattern: "3-tier customer-facing analytics",
  routing_mode: "per-customer database namespace",
  customers: {
    acme: summarizeCustomer(customerData.acme),
    globex: summarizeCustomer(customerData.globex),
  },
};

console.log(JSON.stringify(result, null, 2));

File: plugins/motherduck/skills/motherduck-build-cfa-app/artifacts/customer_routing_example.py
import json
import sys
from pathlib import Path

import duckdb

sys.path.append(str(Path(__file__).resolve().parents[3]))

from scripts._lib.motherduck_artifact_utils import artifact_session


def fetch_rows(conn: duckdb.DuckDBPyConnection, sql: str) -> list[dict]:
    cursor = conn.execute(sql)
    columns = [col[0] for col in cursor.description]
    return [dict(zip(columns, row)) for row in cursor.fetchall()]


def main() -> None:
    with artifact_session(
        slug="motherduck-build-cfa-app",
        database_keys=["customer_acme", "customer_globex"],
    ) as session:
        conn = session.conn
        for db_key, values in {
            "customer_acme": [(1, "search", 12.5), (2, "checkout", 18.0)],
            "customer_globex": [(1, "signup", 4.0), (2, "invoice_paid", 9.5)],
        }.items():
            conn.execute(
                f"""
                CREATE TABLE {session.table(db_key, "main", "analytics_events")} (
                    event_id INTEGER,
                    event_type VARCHAR,
                    revenue DOUBLE
                )
                """
            )
            conn.executemany(
                f"INSERT INTO {session.table(db_key, 'main', 'analytics_events')} VALUES (?, ?, ?)",
                values,
            )

        def query_customer(database_key: str) -> list[dict]:
            return fetch_rows(
                conn,
                f"""
                SELECT event_type, SUM(revenue) AS total_revenue
                FROM {session.table(database_key, "main", "analytics_events")}
                GROUP BY 1
                ORDER BY total_revenue DESC
                """,
            )

        result = {
            "backend": session.describe(),
            "pattern": "3-tier customer-facing analytics",
            "routing_mode": "per-customer database namespace",
            "customers": {
                "acme": query_customer("customer_acme"),
                "globex": query_customer("customer_globex"),
            },
        }
        print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()
